### PR TITLE
Jt/net 81 update interactalpha

### PR DESCRIPTION
**Pull request by jterzis**
NET-81

new InteractAlphaSparse script that 
1. reads from json file output by bytecode-diff command
2. removes facets corresponding to diamonds that have changed as evidenced in json output
3. redeploys those facets with diamondCuts to the right diamonds

Full pipeline to update facets that have changed since last time the report was generated:

Generate source diff report comparing hashes of facet bytecode emitted from forge build against previous invocations (stored in reports in s3 or locally)
1. `GOWORK=off go run main.go -s --source-diff-log s3://source-diff-reports -v`
2. Generate new source-diff report
```
cd scripts/bytecode-diff
mkdir source-diffs && cd source-diffs
# use custom profile or aws env var credentials to cp remote source diff report from s3 and convert to json
aws s3 cp s3://source-diff-reports/$(aws s3 ls s3://source-diff-reports --profile river-s3-access | sort | tail -n 1 | awk '{print $4}') ./compiled_source_diff.yaml --profile river-s3-access && yq -o=json ./compiled_source_diff.yaml > compiled_source_diff.json
```

```
# print key vals from json report. Report is split into "existing", "updated" facet arrays. We want to deploy only facets from updated facet array.

➜  river git:(jt/net-81-update-interactalpha) ✗ jq '.existing[] | select(.diamond == "baseRegistry")' compiled_source_diff.json
{
  "diamond": "baseRegistry",
  "facets": [
    {
      "facetName": "DiamondCutFacet",
      "deployedAddress": "0x3cE1e78Ce5c1009A494Cd2Bae3E10e5ac3a3EFB9",
      "sourceHash": "0xcee277749bc5362b5860c0b75c57605c38296f41c02e6023fb51ccfb6aa7a03d"
    },
    {
      "facetName": "DiamondLoupeFacet",
      "deployedAddress": "0xf8c7538bd41465dc46986dC9CE9AbfD4DF673d3b",
      "sourceHash": "0x5655d34546179992f27e3e521a9657cdc4b197d25fd55c480dba46682be8d61c"
    },
    {
      "facetName": "OwnableFacet",
      "deployedAddress": "0x7287db3ec24092EBFf9AA3e3B39D8da737b96E13",
      "sourceHash": "0x118754225c699bab1d584814cc1da45e9fa34b4e129dc82239c5abfd29bfb820"
    },
    {
      "facetName": "ERC721ANonTransferable",
      "deployedAddress": "0x5cC053a7aF99cCEC1Fc969362fDeEFe13CA77503",
      "sourceHash": "0x19df8ad32ac730aa50fef283fad7ccf52c9d13c76d13ca24ea0d16a4a9555559"
    },
    {
      "facetName": "NodeOperatorFacet",
      "deployedAddress": "0xF032CE2426C202437bfea0FE2f0125d5cD242C97",
      "sourceHash": "0xa6066c4de54a34d75536071551a56e8fde94efa2ae59b0ceed2e6875592cccd5"
    },
    {
      "facetName": "MetadataFacet",
      "deployedAddress": "0x1C691cec67d67afDFA39BbAAFF3896B64cE32759",
      "sourceHash": "0xbaf018043710b7d8837d853efd1b0f95047b3045db6d5040864eebdc8044a706"
    },
    {
      "facetName": "EntitlementChecker",
      "deployedAddress": "0xf65A8d463B1A404Ae98CdD0352Afb786C5FE8902",
      "sourceHash": "0xaaec24a28692c2e28aa3af58877fe1539ff07b6535b392d9b99e86224cc9564a"
    },
    {
      "facetName": "RewardsDistribution",
      "deployedAddress": "0xc464871dbe3E342577Dd7530Fa74F99C9423E173",
      "sourceHash": "0x0e7a2fa2097d4edad9cb072e8f52f4f8216399d4522cd849491e32bb90cfeb4f"
    },
    {
      "facetName": "SpaceDelegationFacet",
      "deployedAddress": "0x041cd06156dbC49f91dF76AC7E9d8c916D466f4C",
      "sourceHash": "0x8961951f23be4523c524cb1cc088437d5764b5ed23c9b7bc5c67a7750770ef75"
    }
  ],
  "numFacets": 9
}
```
3. Run interaction script to update facets changed since last report

```
make interact-any-explicit \
                  context=alpha \
                  rpc=base_sepolia \
                  contract=InteractAlphaSparse \
                  private_key=${{ env.ALPHA_BASE_SEPOLIA_DEPLOYER_PK }} \
                  etherscan=${{ env.BLOCKSCOUT_SEPOLIA_API_KEY }} \
                  verifier=${{ env.BLOCKSCOUT_SEPOLIA_URL }}
```

### CI update plan
Once this is pulled into harmony, `Update_alpha_facets.yml` will be updated to run this interaction script as part of daily alpha ci job.

### CODE REVIEW

File: contracts/scripts/deployments/diamonds/DeploySpace.s.sol [hunk 1]
```diff
@@ -189,6 +189,115 @@ contract DeploySpace is DiamondHelper, Deployer {
...
+          entitlementGatedHelper.makeCut(
+            entitlementGated,
+            IDiamond.FacetCutAction.Add
+          )
+        );
+      }
+    }
+
+    return
+      Diamond.InitParams({
```
Comment (giuseppecrj) [inline contracts/scripts/deployments/diamonds/DeploySpace.s.sol:290]: don't need to return InitParams

File: contracts/scripts/interactions/InteractAlphaSparse.s.sol [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        facetNames[j] = diamonds[i].facets[j].facetName;
+      }
+
+      bytes32 diamondNameHash = keccak256(abi.encodePacked(diamondName));
+
+      if (diamondNameHash == keccak256(abi.encodePacked("space"))) {
+        // deploy space diamond by facets
+        diamondAddress = getDeployment("space");
+        removeRemoteFacetsByAddresses(deployer, diamondAddress, facetAddresses);
+        deploySpace.diamondInitParamsFromFacets(deployer, facetNames);
```
Comment (giuseppecrj) [inline contracts/scripts/interactions/InteractAlphaSparse.s.sol:142]: deploySpace.addCut(facetName)

**Review by giuseppecrj — CHANGES_REQUESTED** [review]

File: contracts/scripts/deployments/diamonds/DeploySpace.s.sol (hunk 1 above)
Comment (jterzis) [reply contracts/scripts/deployments/diamonds/DeploySpace.s.sol:290]: 👍

File: contracts/scripts/interactions/InteractAlphaSparse.s.sol (hunk 2 above)
Comment (jterzis) [reply contracts/scripts/interactions/InteractAlphaSparse.s.sol:142]: addCut is called for each facet in facetNames within the body of diamondInitParamsFromFacets

File: contracts/scripts/deployments/diamonds/DeploySpace.s.sol (hunk 1 above)
Comment (jterzis) [reply contracts/scripts/deployments/diamonds/DeploySpace.s.sol]: 76e0b6ce6d8744d1185cee3a3e1cc45acad5f03a

**Review by giuseppecrj — APPROVED** [review]
Seems like a good first step, I think we should properly build a better system to avoid the if/else statements where we save the helper by string and the initData by facet

**Pull request merged**